### PR TITLE
feat(ctb): `PrestateRegistry`

### PR DESCRIPTION
**Pull request by clabby**
## Overview

Adds the new `PrestateRegistry` contract, which contains hardfork definitions as well as a mapping of prestates, agnostic to various fault proof virtual machines and programs.

### Semantics

The `PrestateRegistry` is governed by the `Guardian` role, and only has a single stateful object:

```solidity
/// @notice Represents a hardfork activation and the prestate for the `op-program` at that hardfork.
/// @custom:field prestates Mapping of `VM ID` -> `Program ID` -> absolute prestate hash.
/// @custom:field l2ActivationBlock The L2 block number at which the hardfork is activated.
struct Hardfork {
    mapping(uint256 => mapping(uint256 => Hash)) prestates;
    uint256 l2ActivationBlock;
}

/// @notice A list of L2 hardforks, in ascending order. This list does not have to be exhaustive, but it must be
///         sorted and contain upgrades that alter the Fault Proof Program prestate hashes.
Hardfork[] public hardforks;
```

The `Guardian` role is allowed to:

1. Append a new `Hardfork` to the list.
2. Revoke a pending hardfork that has not yet activated.
   i. _NOTE:_ There is no perfect way to do this, atm we are comparing the L2 timestamp with the L1 timestamp to check if the fork has already activated. While this isn't precise, there should only be a small drift, which should be fine for what we need it for.

**`activePrestate`**

```solidity
/// @notice Returns the absolute prestate in tthe latest active hardfork for a given VM and program.
/// @param _l2BlockNumber The L2 block number at which the program is being ran until.
/// @param _vmID The VM ID.
/// @param _programID The program ID.
function activePrestate(
    uint256 _l2BlockNumber,
    uint256 _vmID,
    uint256 _programID
)
    public
    view
    returns (Hash _prestate);
```

The registry's core function is `activePrestate`, which returns the absolute prestate hash active at a given L2 block number for a VM/program ID pair. This will be consumed by the `FaultDisputeGame` to determine the absolute prestate used during initialization.

### Integration

This PR does not integrate the `PrestateRegistry` into the `FaultDisputeGame` or `MIPS` VM just yet, which will be added into a separate PR. Each FDG will assign a program ID,
and each FPVM will assign a VM ID through the `IBigStepper` interface. With these values, the FDG will query the `PrestateRegistry` in the `initialize` function, grabbing the
active prestate hash for the L2 block number that the output proposal corresponds to.

### CODE REVIEW

**Comment by clabby** [comment]
* **#9751** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/9751?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `develop`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @clabby and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/contracts-bedrock/src/dispute/PrestateRegistry.sol [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        // INVARIANT: If the fork has already been activated, it cannot be revoked.
+        // NOTE: This check is imperfect as we reference based on L1 time, but this drift is acceptable.
+        Hardfork storage latestFork = hardforks[hardforks.length - 1];
+        if (l2BlockToTimestamp(latestFork.l2ActivationBlock).raw() <= block.timestamp) {
+            revert ForkAlreadyActivated();
+        }
+
+        // Remove the latest hardfork.
+        hardforks.pop();
+    }
```
[the remark this answers is not in the document]
Comment (clabby) [reply packages/contracts-bedrock/src/dispute/PrestateRegistry.sol]: 👍

File: packages/contracts-bedrock/src/dispute/PrestateRegistry.sol [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    /// @custom:field vmID The VM ID.
+    /// @custom:field programID The program ID.
+    /// @custom:field prestateHash The prestate hash for the program on the given VM.
+    struct PrestateInformation {
+        uint256 vmID;
+        uint256 programID;
+        Hash prestateHash;
+    }
+
+    /// @notice The superchian configuration.
```
Comment (smartcontracts) [inline packages/contracts-bedrock/src/dispute/PrestateRegistry.sol]:
```suggestion
    /// @notice The superchain configuration.
```

File: packages/contracts-bedrock/src/dispute/PrestateRegistry.sol [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    SuperchainConfig internal immutable SUPERCHAIN_CONFIG;
+    /// @notice The L2 genesis block timestamp.
+    Timestamp internal immutable L2_GENESIS_BLOCK_TIMESTAMP;
+    /// @notice The L2 genesis block.
+    uint256 internal immutable L2_GENESIS_BLOCK;
```
Comment (smartcontracts) [inline packages/contracts-bedrock/src/dispute/PrestateRegistry.sol]: L2 genesis block number?

File: packages/contracts-bedrock/src/dispute/PrestateRegistry.sol [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+    /// @notice Revokes the latest hardfork, if it has not happened yet.
+    function revokePendingFork() external {
+        // INVARIANT: Only the guardian can revoke pending hardforks.
+        if (msg.sender != SUPERCHAIN_CONFIG.guardian()) revert BadAuth();
+
+        // INVARIANT: If the fork has already been activated, it cannot be revoked.
+        // NOTE: This check is imperfect as we reference based on L1 time, but this drift is acceptable.
+        Hardfork storage latestFork = hardforks[hardforks.length - 1];
+        if (l2BlockToTimestamp(latestFork.l2ActivationBlock).raw() <= block.timestamp) {
```
Comment (smartcontracts) [inline packages/contracts-bedrock/src/dispute/PrestateRegistry.sol:97]: Consider adding a buffer here, to account for the drift. Depends on the timestamp drift that the L2 allows though.

File: packages/contracts-bedrock/src/dispute/PrestateRegistry.sol [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+            revert ForkAlreadyActivated();
+        }
+
+        // Remove the latest hardfork.
+        hardforks.pop();
+    }
+
+    /// @notice Returns the absolute prestate in tthe latest active hardfork for a given VM and program.
```
Comment (smartcontracts) [inline packages/contracts-bedrock/src/dispute/PrestateRegistry.sol]:
```suggestion
    /// @notice Returns the absolute prestate in the latest active hardfork for a given VM and program.
```

File: packages/contracts-bedrock/src/dispute/PrestateRegistry.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (clabby) [reply packages/contracts-bedrock/src/dispute/PrestateRegistry.sol]: I think this is fine tbh @smartcontracts. It's an expectation that if a hardfork is deleted, the one that takes its place will override the legacy prestate values again in `registerHardfork`. wdyt?

File: packages/contracts-bedrock/src/dispute/PrestateRegistry.sol (hunk 4 above)
Comment (clabby) [reply packages/contracts-bedrock/src/dispute/PrestateRegistry.sol:97]: I think it's tight enough that we probably don't need a buffer. This doesn't need to be perfectly precise, and we'd likely not revoke a fork minutes before it happened.

**Pull request closed without merging**